src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -7,504 +7,951 @@ saturation specific humidities
+    return T_1
+  else  # If saturated, iterate
+    T_2 = air_temperature_from_liquid_ice_pottemp(θ_liq_ice, p, q_tot, DT(0), q_tot) # Assume all ice
+    eos_root(T, args) = θ_liq_ice - liquid_ice_pottemp_sat(T, p, q_tot, phase_partitioning_eq(T, ρ, q_tot)...)
+    T, converged = find_zero(eos_root, T_1, T_2, Tuple(1),
+                             IterParams(DT(1e-3), 10),
+                             SecantMethod())
+    return T
   end
 end
 
 """
     liquid_ice_pottemp(T, p[, q_tot=0, q_liq=0, q_ice=0])
 
-Return the liquid-ice potential temperature, given the temperature `T`,
-pressure `p`, total specific humidity `q_tot`, liquid specific humidity `q_liq`,
-and ice specific humidity `q_ice`.
+The liquid-ice potential temperature where
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
 """
-function liquid_ice_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0)
-
+function liquid_ice_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
     # isobaric specific heat of moist air
     _cp_m   = cp_m(q_tot, q_liq, q_ice)
 
     # liquid-ice potential temperature, approximating latent heats
     # of phase transitions as constants
-    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(LH_v0*q_liq + LH_s0*q_ice)/(_cp_m*T))
+    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * (1 - (DT(LH_v0)*q_liq + DT(LH_s0)*q_ice)/(_cp_m*T))
+end
+
+"""
+    liquid_ice_pottemp(ts::ThermodynamicState)
+
+The liquid-ice potential temperature,
+given a thermodynamic state `ts`.
+"""
+liquid_ice_pottemp(ts::ThermodynamicState) =
+  liquid_ice_pottemp(air_temperature(ts), air_pressure(ts), ts.q_tot,
+                     phase_partitioning_eq(ts)...)
+
+"""
+    dry_pottemp(T, p, q_tot, q_liq, q_ice)
+
+The dry potential temperature where
+
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+ """
+dry_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where {DT} =
+  T / exner(p, q_tot, q_liq, q_ice)
+
+"""
+    dry_pottemp(ts::ThermodynamicState)
+
+The dry potential temperature, given a thermodynamic state `ts`.
+"""
+dry_pottemp(ts::ThermodynamicState) =
+  dry_pottemp(air_temperature(ts), air_pressure(ts), ts.q_tot,
+              phase_partitioning_eq(ts)...)
+
+"""
+    air_temperature_from_liquid_ice_pottemp(θ, p[, q_tot=0, q_liq=0, q_ice=0])
+
+The air temperature, where
+
+ - `θ` potential temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
 
+Without the specific humidity arguments, the results
+are that of dry air.
+"""
+function air_temperature_from_liquid_ice_pottemp(θ::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
+    return θ*exner(p, q_tot, q_liq, q_ice)
```

Comment:
Here you need to return this: `θ*exner(p, q_tot, q_liq, q_ice) + (DT(LH_v0)*q_liq + DT(LH_s0)*q_ice)/_cp_m` with the usual definition for `_cp_m`. I would also replace `θ` by `θ_liqice` or something like it to make clear it is the liquid ice potential temperature.

Reply:
Thank you!